driver-core/src/main/com/mongodb/internal/operation/CursorResourceManager.java

```diff
@@ -182,7 +182,7 @@ void endOperation() {
     void close() {
         boolean doClose = withLock(lock, () -> {
             State localState = state;
-            if (localState == State.OPERATION_IN_PROGRESS) {
+            if (localState.inProgress()) {
```

Comment:
The problem with the original code:

If `AsyncCommandBatchCursor.close` is called concurrently more than once while `AsyncCommandBatchCursor.next` is running (not necessarily the `next` method itself, but the asynchronous logic of this method), then the following state transitions were possible:

1. One of the `close` calls observes `State.OPERATION_IN_PROGRESS`, and changes it to `State.CLOSE_PENDING`. So far so good.
2. The other `close` observes `State.CLOSE_PENDING` and changes it to `State.CLOSED`, which breaks one of the invariant that are supposed to be maintained by `CursorResourceManager`: while an operation is in progress, `state.inProgress()` must be `true`, i.e., `state` must be either `OPERATION_IN_PROGRESS` or `CLOSE_PENDING` (implies `OPERATION_IN_PROGRESS`).